File: fixtures/crash/crash.go
// crash.go - Go counterpart of crash.cpp.
//
// Go has no abort(); a nil dereference and an integer divide-by-zero both
// raise a runtime panic that surfaces as a fatal signal to the debugger.
package main

import (
	"fmt"
	"os"
	"strings"
)

func doSegv() {
	var p *int
	*p = 42 // @dap:before_crash
}

func doDivzero(z int) {
	x := 1
	y := x / z // @dap:before_crash_div
	fmt.Println(y)
}

func main() {
	mode := "segv"
	for _, a := range os.Args[1:] {
		if strings.HasPrefix(a, "--mode=") {
			mode = strings.TrimPrefix(a, "--mode=")
		}
	}
	fmt.Printf("crash mode=%s\n", mode)

	switch mode {
	case "divzero":
		doDivzero(0)
	default:
		doSegv()
	}
}

File: fixtures/crash/crash.rs
// crash.rs - Rust counterpart of crash.cpp.
//
// Note Rust's integer divide-by-zero is a runtime panic (not SIGFPE), and
// `abort` calls std::process::abort(); both produce a debugger stop.
use std::env;

fn do_segv() {
    let p: *mut i32 = std::ptr::null_mut();
    unsafe {
        *p = 42; // @dap:before_crash
    }
}

fn do_abort() {
    std::process::abort(); // @dap:before_crash_abort
}

fn do_divzero(z: i32) {
    let x = 1;
    let y = std::hint::black_box(x) / std::hint::black_box(z); // @dap:before_crash_div
    println!("{}", y);
}

fn main() {
    let mode = env::args()
        .skip(1)
        .find_map(|a| a.strip_prefix("--mode=").map(str::to_owned))
        .unwrap_or_else(|| "segv".to_owned());
    println!("crash mode={}", mode);

    match mode.as_str() {
        "abort" => do_abort(),
        "divzero" => do_divzero(0),
        _ => do_segv(),
    }
}
